### PR TITLE
Implement credentials UI for campaigns user tokens

### LINKED ISSUE #14990: campaigns: add user token settings interface
We'll need GraphQL queries/mutations to support this, and eventually some design input.

### Objectives

* [ ] A GraphQL API must be available to manage tokens.
* [ ] Users should be able to paste in tokens provisioned from their code host and have them be used.
* [ ] Users should be able to delete tokens from the Sourcegraph database.
* [ ] Users should be able to see which code hosts they have tokens for and which they do not.

### Security considerations

* User tokens *must not* be returned from the GraphQL API at any point.
* Listing code hosts that the user may add tokens for must be filtered by the repositories that the user has access to.

### PR DESCRIPTION
**Pull request by eseliger**
This adds an implementation of the GraphQL API resolvers and the UI part of the campaigns user tokens project.

Closes #14990

### CODE REVIEW

File: enterprise/internal/campaigns/resolvers/resolver.go [hunk 1]
```diff
@@ -656,6 +710,97 @@ func (r *Resolver) SyncChangeset(ctx context.Context, args *graphqlbackend.SyncC
...
+		ExternalServiceType: extsvc.KindToType(kind),
+		UserID:              actor.UID,
+	}
+
+	// Throw error documented in schema.graphql.
+	_, err = db.UserCredentials.GetByScope(ctx, scope)
+	if err != nil && !errcode.IsNotFound(err) {
+		return nil, err
+	}
+	if !errcode.IsNotFound(err) {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/resolver.go:752]: This should probably check whether the returned result is not nil.

File: enterprise/internal/campaigns/store_codehost.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ORDER BY external_service_type ASC, external_service_id ASC
+`
+
+func listCodeHostsQuery() *sqlf.Query {
+	preds := []*sqlf.Query{
+		// Only show supported hosts.
+		sqlf.Sprintf("external_service_type IN (%s, %s, %s)", extsvc.TypeGitHub, extsvc.TypeGitLab, extsvc.TypeBitbucketServer),
+		// And only for those which have any enabled repositories.
+		sqlf.Sprintf("repo.deleted_at IS NULL"),
+	}
```
Comment (mrnugget) [inline enterprise/internal/campaigns/store_codehost.go:43]:
```suggestion
	preds := []*sqlf.Query{
		// And only for those which have any enabled repositories.
		sqlf.Sprintf("repo.deleted_at IS NULL"),
	}
	supportedTypes := []*sqlf.Query{}
        for extSvcType := range campaigns.SupportedExternalServices {
		supportedTypes = append(supportedTypes, sqlf.Sprintf(extSvcType))
	}

	preds = append(preds, sqlf.Sprintf("external_service_type IN (%s)", sqlf.Join(supportedTypes, ", "))
```

**Review by mrnugget — APPROVED** [review]

File: enterprise/internal/campaigns/resolvers/code_host_connection.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		return nil, err
+	}
+
+	credsByIDType := make(map[idType]*db.UserCredential)
+	for _, cred := range creds {
+		t := idType{
+			externalServiceID:   cred.ExternalServiceID,
+			externalServiceType: cred.ExternalServiceType,
+		}
+		credsByIDType[t] = cred
```
[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/campaigns/resolvers/code_host_connection.go:69]: No, they are having a unique constraint in the database.

File: client/web/src/enterprise/campaigns/settings/AddCredentialModal.tsx [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    afterCreate: () => void
+    history: H.History
+    externalServiceKind: ExternalServiceKind
+    externalServiceURL: string
+}
+
+const modalTitles: Record<ExternalServiceKind, string> = {
+    [ExternalServiceKind.GITHUB]: 'GitHub',
+    [ExternalServiceKind.GITLAB]: 'GitLab',
+    [ExternalServiceKind.BITBUCKETSERVER]: 'Bitbucket Server',
```
[the remark this answers is not in the document]
Comment (eseliger) [reply client/web/src/enterprise/campaigns/settings/AddCredentialModal.tsx:23]: Working on a spike now via Zoom.

**Pull request merged**